WIP resolves Issue 3884: extension documentation

djencks opened the pull request:
This has most of the basic structure and much of the content I'd like to see in the extensions documentation. Note that so far among the processor pages only block-macro-processor  is filled in.

djencks commented:
This is now ready for review.

danyill reviewed (commented):
What a great job -- there's a lot of really good information here and I learnt a lot.

I'm not the best reviewer for this but I've left a a few comments, hopefully which may be of use...
I'll keep reviewing.

danyill commented on docs/modules/extensions/pages/processors/tree-processor.adoc:
```diff
@@ -1,5 +1,38 @@
+
+A block found this way can be replaced with `parent.blocks[parent.blocks.index original_block] = new_block`.
+
+=== Traverse the block tree
+
+The code here is more complicated and needs to be recursive to find all blocks, but the attributes present at each block may be computed as the traversal proceeds.
+To replay the attributes, for each block execute `document.playback_attributes block.attributes`.
+Remember to call `document.restore_attributes` before returning.
```
I think it might be good to have an example of this -- is this so you can determine the state of an attribute at an arbitrary position in the document?

djencks replied:
I think you're right about needing an example.

mojavelinux replied:
The subject of tree traversal deserves its own page. It's an important topic, but also an advanced one. That page can set up the scenario when tree traversal is needed and go over the caveats of doing it. It's not something that is really part of the standard Asciidoctor usage and you need to know that you are offroading it a bit (at least as it stands now).

djencks replied:
I've been looking for examples of when tree traversal is needed and the only one I can come up with is fixing STEM processing, which ought to be doable with a block processor rather than a tree processor anyway. Deemphasizing tree traversal is definitely a good idea.

danyill commented on docs/modules/extensions/pages/processors/tree-processor.adoc:
```diff
@@ -1,5 +1,38 @@
+
+=== Query the document blocks using the `document.find_by` method
+
+This provides quick simple access to the blocks of interest, but does not provide access to the attributes set inline in the document.
```
```suggestion
This provides quick simple access to the blocks of interest, but does not provide access to the attributes set inside in the document.
```

We can't set attributes inline can we? I think we're talking about setting attributes outside the document header, is that right?

djencks replied:
"set in the document after the document header"?

mojavelinux replied:
> but does not provide access to the attributes set inside in the document.

This should be framed as "does not honor attributes that are set in the body of the document".

djencks commented on docs/modules/extensions/pages/index.adoc:
```diff
@@ -1,51 +1,76 @@
 
-Asciidoctor also allows extensions to be written using the full power of a programming language (whether it be Ruby, Java, Groovy or JavaScript).
-You don't have to shave yaks to get the functionality you want, and you can distribute the extension using defacto-standard packaging mechanisms like RubyGems or JARs.
+Asciidoctor also allows extensions to be written using the full power of the programming language of your asciidoctor variant: Ruby, Java or other JVM based languages, or Javascript.
```
yes.

djencks commented on docs/modules/extensions/pages/index.adoc:
```diff
@@ -1,51 +1,76 @@
+See xref:processors/block-processor.adoc[].
 
-Compound block processor::
+Compound block processor:::
```
Actually this item doesn't belong in this list at all, even if we keep it as an example.

djencks commented on docs/modules/extensions/pages/index.adoc:
```diff
@@ -1,51 +1,76 @@
 
-Inline macro processor::
+Inline macro processor (Syntax)::
 Registers a custom inline macro and processes it (e.g., `btn:[Save]`).
```
Thanks for the catch!
I thought I explained the regex feature on the inline macro page.  However, I haven't used it myself and I might not understand its power.  Can you give an example of what you're describing here?

djencks replied:
We have Danyill's nice example on the inline macro page, so closing.

djencks commented on docs/modules/extensions/pages/index.adoc:
```diff
@@ -1,51 +1,76 @@
 
-Compound block processor::
+Compound block processor:::
 Register a custom block named `collapsible` that transforms a listing block into a compound block.
```
dropped :-)

djencks commented on docs/modules/extensions/pages/packaging.adoc:
```diff
@@ -0,0 +1,62 @@
+= Packaging an extension
+
+Fundamentally this involves making the extension code visible to asciidoctor and registering the processors globally or with a particular registy.
```
I'm changing regist[r]y to Registry instance, WDYT?

djencks commented on docs/modules/extensions/pages/index.adoc:
```diff
@@ -1,51 +1,76 @@
-Processes the `include::<filename>[]` directive.
-See xref:include-processor.adoc[].
+Docinfo Processor (Document)::
+Adds additional content to the header or footer regions of the generated document.
```
I think this should go on the docinfo processor page.

djencks commented on docs/modules/extensions/pages/block-example.adoc:
```diff
@@ -0,0 +1,109 @@
+include::example$open-block-extension-class.rb[]
+----
+
+Until nested open blocks become part of the Asciidoc language, this is actually a useful extension for html.
+It allows you to lay out your content in arbitrarily nested `div` elements, with specified css classes (mapped from the block role).
```
Yes. I reordered the sentences and added a link to the issue.

djencks commented on docs/modules/extensions/pages/block-example.adoc:
```diff
@@ -0,0 +1,109 @@
+
+== Extension as a block
+
+Here's the 'same' extension expressed as a block:
```
same functionality expressed as a block?

djencks commented on docs/modules/extensions/pages/register.adoc:
```diff
@@ -1,26 +1,135 @@
+
+Code block::
+The block will be called by a Registry at an appropriate time to inform the Registry of the processors.
+Note that if the block takes no parameters, or a block (arity 0 or -1), the registry will use `instance_exec` to call the block, so the Registry methods such as `inlineMacro` are available directly in the block.
```
Not surprising.  I've reworded extensively.

djencks commented on docs/modules/extensions/pages/register.adoc:
```diff
@@ -1,26 +1,135 @@
 ----
-Asciidoctor::Extensions.register do |document|
+      registry = Asciidoctor::Extensions.create
+      registry.tree_processor SampleTreeProcessor //We are supplying a processor class
```
thanks!

djencks commented on docs/modules/extensions/pages/register.adoc:
```diff
@@ -1,26 +1,135 @@
+= Register Processors
 
-== Register one or more extensions
+There are several terms that it's easy to mix up.
```
I'm trying moving the definitions to a glossary page.
Your wording isn't all that accurate, but something is needed.... take a look at what I came up with.

djencks commented on docs/modules/extensions/pages/block-example.adoc:
```diff
@@ -0,0 +1,109 @@
+Running (in the directory of the open-block-extension file)
+[source,shell]
+----
+echo '[openblock,foo]\n----\nInside the Open Block\n----' | asciidoctor -s -r ./open-block-extension.rb -
```
I think I fixed this :-)

djencks commented on docs/modules/extensions/pages/register.adoc:
```diff
@@ -1,26 +1,135 @@
+The most direct method is to explicity create a registry using `Asciidoctor::Extensions.create`, directly add processors to that registry, and supply the registry to a document as `extension_registry: my_registry`.
+For instance:
 
 [source,ruby]
```
This is an advanced technique that just about no one will need to use, so I don't think an example is needed.
The idea is to show what happens when you use a group.

djencks commented on docs/modules/extensions/pages/register.adoc:
```diff
@@ -1,26 +1,135 @@
+
+=== Groups
+
+More usually registration code is organized into one or more `Group`.
```
I'm going with 'instances' instead of 'objects' because usually they are blocks.

djencks commented on docs/modules/extensions/pages/block-example.adoc:
```diff
@@ -0,0 +1,109 @@
+= Block extension example
```
I'm thinking about this.... so far I like my example.  Possibly it can be made simpler, I think removing the 'role' attribute is unnecessary.

djencks replied:
The code now could not be simpler IMO :-).  Even if you don't understand why css classes might be interesting, you can see in the output that the blocks turn into "openblock". Since all the examples are html I think that's sufficient. WDYT?

danyill replied:
Sure, looks fine.

mojavelinux replied:
I also think that the opening example should be the shout block. It is the simplest functional example I could come up with and I think it gives the user a quick win. I'm fine with advancing on to the nestable block example right after that. You could say something like "with that under your belt, let's look at a slightly more complex example". This is a very typical progression. But we can't just throw the reader into the deep end right out of the gate.

djencks replied:
Compared with the shout block, the nestable block is fewer lines of code, simpler, and useful. How is this the deep end?  And how mixed are your metaphors :-)? However, this isn't worth arguing about.

djencks commented on docs/modules/extensions/pages/index.adoc:
```diff
@@ -1,51 +1,76 @@
-Asciidoctor provides the following extension points:
+Document Processor::
+  A document processor does not have an explicit invocation in the document.
+Except for `IncludeProcessor`, a document processor is invoked at the appropriate stage of processing with the entire document.
```
I'm going to experiment to see if there's a way to interleave a description of the processing stages and the processor types.  It may take a while :-)

djencks replied:
I'm reasonably happy with what I came up with for an interleaved description.

danyill replied:
Yes, it looks really great, much better than my attempts :grinning:

djencks replied:
I agree we're not there yet.  I think the crucial distinctions are that a doc processor operates on text, and has no invocation, and a syntax processor operates on the DOM/AST tree.

djencks commented on docs/modules/extensions/pages/index.adoc:
```diff
@@ -1,51 +1,76 @@
+See xref:processors/block-macro-processor.adoc[].
+
+Tree processor (Document)::
+Processes the [.class]#Asciidoctor::Document# (AST) once parsing is complete.
```
I think asking Dan about this is a good idea...

djencks commented on docs/modules/extensions/pages/register.adoc:
```diff
@@ -1,26 +1,135 @@
+== Adding processors to a Registry
+
+Processors are added to a Registry by calling a registry method named after the type of processor.
+For instance, to add a block macro, call `registry.blockMacro ...`.
```
Well, the document processor method names all end in processor and the syntax ones don't, but I'm going to try mentioning the registration methods on the processor specific pages. More detail may be needed...

danyill replied:
LGTM

mojavelinux replied:
@djencks is exactly right. The DSL methods for the syntax processors do not include the term processor because the focus is on the syntax (block, block macro, inline macro). The lifecycle processors need this, otherwise it just doesn't read we'll (pre, tree, post). So I kept the processor term there one so it reads better, but also to distinguish them from syntax processors.

Mogztter commented on docs/modules/extensions/pages/index.adoc:
```diff
@@ -22,11 +22,6 @@ Asciidoctor provides the following extension points, shown in the order of invoc
-Since inline content is not parsed until the convert phase, the inline macro processor must return converted text (e.g., HTML) rather than an AST node.
-Once Asciidoctor is changed to https://github.com/asciidoctor/asciidoctor/issues/61[process inline content during the parse phase^], the inline macro processor will need to return an inline node.
-When that switch occurs, there will either be some sort of adapter or required migration for inline macro processors, but that has yet to be determined.
-
```
I think we should keep this note until #61 is fixed.
You can already return an inline node (instead of a string) but the rest of the text is still accurate.

Mogztter commented on docs/modules/extensions/pages/processors/inline-macro-processor.adoc:
```diff
@@ -3,6 +3,15 @@
 
 InlineMacroProcessor is a SyntaxProcessor
 
+Note:
```
No admonition?

djencks replied:
Fixed!

Mogztter commented on docs/modules/extensions/pages/processors/inline-macro-processor.adoc:
```diff
@@ -3,6 +3,15 @@
+Note:
+Currently, uses of inline macros are recognized and processed during the conversion phase of a parsed block.
+Furthermore, uses are recognized by means of a regular expression.
+One of the goals of many participating the in Asciidoc language standardization effort is to sufficiently define the language so that inline macro uses can be recognized and processed during the parse phase, so the inline Nodes returned from processing form part of the DOM or parse tree.
```
```suggestion
One of the goals of many participating in the AsciiDoc language standardization effort is to sufficiently define the language so that inline macro uses can be recognized and processed during the parse phase, so the inline Nodes returned from processing form part of the DOM or parse tree.
```

I think this sentence is a bit long. Maybe we should use a dot and not a comma?
I also think that we should expand the acronym DOM and probably state that we are not talking about HTML DOM.

djencks replied:
Yes. Rewritten a bit.

Mogztter commented on docs/modules/api/pages/options.adoc:
```diff
@@ -53,13 +55,13 @@ _(Experimental)._
-|Overrides the extensions registry instance.
-Instead of providing a Ruby block containing extensions to register, this option lets you replace the extension registry itself, giving you complete control over how extensions are registered for this processor.
+|Specifies an explicit xref:extensions:glossary.adoc#id_registry[extensions registry] instance.
+Instead of providing a processor extension group to register extensions, using the :extensions option, this option lets you replace the extension registry itself, giving you complete control over how extensions are registered for this processor.
```
> using the :extensions option, this option lets...

We should probably try to avoid this repetition

djencks replied:
Rewritten extensively.

mojavelinux commented:
First, I want to say thank you for taking this on because this is not an easy task / subject.

Right now, I think this PR is too big to review. It needs to be split it up. It definitely shouldn't include a new test for the processor. And it seems to contain edits to docs that are only tangentially related to extensions. Let's process those 1-by-1 to trim this down.

I'm also not a fan of the use of a pipe to run the examples. That's a very abnormal usage pattern and we don't want to encourage it. Let's create separate example files and instruct the user to run them. (Just to be clear, I'm find with putting the AsciiDoc in the examples directory). I also don't want to include the output of the example in the docs. That's just going to be a pain to maintain. We should either leave that up to the reader to observe on their on machine or we can find a way to auto-generate them at build time (perhaps using some sort of extension).

I'll try my best to provide feedback on the new written material.

mojavelinux commented on docs/modules/extensions/pages/processors/block-processor.adoc:
```diff
@@ -0,0 +1,135 @@
+| `:comment`
+| `////`
+
+| `:fenced_code`
```
The fenced_code context is obsolete.

djencks replied:
Since this isn't documented anywhere I can find, I'm leaving a comment noting obsolescence in place.

mojavelinux commented on docs/modules/extensions/pages/processors/block-processor.adoc:
```diff
@@ -0,0 +1,135 @@
+Sets the contexts allowed for this block.
+The choices are:
+
+[cols=2,opts=headers]
```
This table should be a referenced rather than defined inline. These are well-defined contexts and they need to be defined elsewhere. They aren't specific to extensions.

djencks replied:
How about allowing it here for now and finding a better place for it later?  I'm not interested in rewriting all of the Asciidoc documentation for now.

mojavelinux commented on docs/modules/extensions/nav.adoc:
```diff
@@ -1,11 +1,14 @@
 * xref:index.adoc[]
```
The document processors should be grouped under "Document Processors" and the syntax processors grouped under "Syntax Processors". Otherwise, it just looks like a laundry list of extension points.

djencks replied:
They are arranged currently in order of invocation.  I've tried adding the type to the page titles, WDYT?  Perhaps having a heading, "processors in order of invocation" would be useful?  If you still want a type/alphabetic arrangement, OK.

mojavelinux commented on docs/modules/extensions/pages/processors/block-macro-processor.adoc:
```diff
@@ -0,0 +1,50 @@
+
+BlockMacroProcessor is a SyntaxProcessor.
+
+== Registration method
```
I like where you are going with this structure, but Registration method is only relevant when defining the extension using a Ruby block. If you define via a class, then we need to specify the class. So perhaps write this as:

```
== Registration

=== DSL method

block_macro

=== Parent class

BlockMacroProcessor
```

or something like that.

djencks replied:
My understanding is that you always need to call the appropriate registry method to add a processor to a registry.
I agree that explicitly mentioning the superclass is a good idea.

mojavelinux commented:
I think there's an important lesson to be learned here in terms of docs management. When we're making a major change to the docs, such as this one, we need to approach that change in stages.

The first stage would be some general organization and cleanup to prepare the docs to be changed. (This might involve extracting content to individual files so they can be included and restructuring the nav). Set the stage, so to speak. Then, the next change would be to introduce the informative pages or sections (aka non-normative docs). The final change would be to rework the reference and examples to be more accurate, in-depth, and actionable.

Trying to do it all at once just creates gridlock during editing. It also does not give the information architect a chance to review the structure before having a pile of content to address.

We need to be more disciplined in order to make the process more efficient for everyone involved and ultimatelyl guarantee a successful outcome.

djencks commented:
...
> We need to be more disciplined in order to make the process more efficient for everyone involved and ultimatelyl guarantee a successful outcome.

Your proposed workflow would be more plausible if you were paying me or there were prompt, say overnight, reviews.  Without that, it's basically a request not to contribute.  I did post my initial thoughts which incorporated a lot of the structure I ended up with about a week before the rest of the major work, with no response.

Without really quick review, I think a reasonable way to approach this is to decide if it is a significant improvement on what is there, and if so commit it, and pursue changes from there.

What if you approached this with the frame of there being no existing extension documentation, and asking if there is anything wrong (as opposed to badly expressed) in the new docs, and anything missing?

djencks commented:
> First, I want to say thank you for taking this on because this is not an easy task / subject.
> 
> Right now, I think this PR is too big to review. It needs to be split it up. It definitely shouldn't include a new test for the processor. And it seems to contain edits to docs that are only tangentially related to extensions. Let's process those 1-by-1 to trim this down.

IIRC the test verifies that a claim in the old documentation is wrong, so it still seems reasonable to me to include it here.  Perhaps I'll look at the other changes you think are tangential.  With the extremely slow PR process for doc changes, there seems to be little point in supplying small PRs, since they tend to go nowhere.
> 
> I'm also not a fan of the use of a pipe to run the examples. That's a very abnormal usage pattern and we don't want to encourage it. Let's create separate example files and instruct the user to run them. (Just to be clear, I'm find with putting the AsciiDoc in the examples directory). I also don't want to include the output of the example in the docs. That's just going to be a pain to maintain. We should either leave that up to the reader to observe on their on machine or we can find a way to auto-generate them at build time (perhaps using some sort of extension).

No problem about the pipe, it's used in a lot of tests and is easy and compact.  The output is already generated by the test script in examples, I guess it could be committed and 'include::'-ed.
> 
> I'll try my best to provide feedback on the new written material.
Thanks, I'm finally looking through it!

djencks commented on docs/modules/extensions/pages/index.adoc:
```diff
@@ -1,51 +1,89 @@
-Asciidoctor also allows extensions to be written using the full power of a programming language (whether it be Ruby, Java, Groovy or JavaScript).
-You don't have to shave yaks to get the functionality you want, and you can distribute the extension using defacto-standard packaging mechanisms like RubyGems or JARs.
+Extensions are central to the power of AsciiDoc because they open up the language to new use cases.
+Asciidoctor provides an extension API that offers a set of extension points.
```
To me, that's exactly what my wording says.  I'll see if I can find a way to make it even clearer.

djencks replied:
How about...
```
Extensions are central to the power of AsciiDoc because they open up the language to new use cases.
To implement this critical functionality, Asciidoctor provides an extension API that offers a set of extension points.
```

djencks commented on docs/modules/extensions/pages/index.adoc:
```diff
@@ -1,51 +1,89 @@
+Extensions in Asciidoctor are easy to write, powerful, and simple to distribute.
+
+Asciidoctor also allows extensions to be written using the full power of the programming language of your Asciidoctor variant: Ruby, Java or other JVM based languages, or Javascript.
+An extension may be included in custom code or packaged and distributed using a language-standard mechanism such as Gems, Jars/Maven, or npm packages.
```
RubyGems and nom packages are both a packaging and distribution mechanism, whereas jars are only a packaging.  To be parallel, Jars/Maven is necessary.

djencks commented on docs/modules/extensions/pages/index.adoc:
```diff
@@ -1,51 +1,89 @@
-Processes the [.class]#Asciidoctor::Document# (AST) once parsing is complete.
-See xref:tree-processor.adoc[].
+The preprocessed lines are streamed into the parse stage.
+The parser converts the lines into a tree of blocks, calling block and block macro processors as their uses are encountered.
```
I think I now understand preprocessors better and their dangers even more so I'll see what I can  come up with.

djencks replied:
I like my text as is.  The details you suggest belong, IMO, on the processor-specific pages.

djencks commented on docs/modules/extensions/pages/register.adoc:
```diff
@@ -1,26 +1,111 @@
+      doc = Asciidoctor::Document.new %(= Document Title\n\ncontent).lines, extension_registry: registry
+----
+
+=== Groups
```
I disagree.  I spent a long time studying the actual code, and IMO the only way to sensibly describe it is as I have done.  Some groups are Groups, but the way the code is set up a conceptual group is the only way to register an extension.  To me, this point of view makes the code comprehensible although very convoluted and using a lot of Ruby magic.  I have no idea why it's written this way, but presenting it in terms of conceptual groups at least makes all the different paths uniform.

djencks commented on docs/modules/extensions/pages/packaging.adoc:
```diff
@@ -0,0 +1,66 @@
+require 'asciidoctor/extensions'
+require 'nestable/processor'
+
+registry = Asciidoctor::Extensions::Registry.new :nestable=> NestableGroup
```
I'll try again, I haven't had much luck making that syntax work despite the alleged support.

djencks commented on docs/modules/extensions/pages/index.adoc:
```diff
@@ -1,51 +1,89 @@
+Asciidoctor provides the following extension points, shown in the order of invocation, interleaved with a description of Asciidoctor processing stages.
+To be more specific, these are types of Processor, and a concrete implementation extends the corresponding Processor subclass.
+
+Asciidoctor starts with a source document and starts reading it into an array of lines and preprocessing it. First, as `include::` directives are processed as they are encountered.
```
List continuations would conceal the interleaved aspect, and section headings would destroy it.  How about making the processing stages a numbered list and the dlist items sublists?

djencks commented on docs/modules/extensions/pages/index.adoc:
```diff
@@ -1,51 +1,89 @@
+This is the actual code that implements the extension functionality.
+The inputs and expected outputs are described on the processor-specific pages linked above.
+
+Capabilities::
```
OK

djencks commented on docs/modules/extensions/pages/index.adoc:
```diff
@@ -1,51 +1,89 @@
+Usually this is done in code that forms a xref:register.adoc#groups[group].
+
+Packaging::
+xref:packaging.adoc[This] involves making the code available to Asciidoctor for the preceding three aspects.
```
OK

djencks commented on docs/modules/extensions/pages/glossary.adoc:
```diff
@@ -0,0 +1,31 @@
+Extension::
+This is an informal usage.
+Some code that extends or modifies Asciidoctor capabilities by means of Processors.
+Processor::
```
I made some changes and this may need more work.

djencks commented on docs/modules/extensions/pages/register.adoc:
```diff
@@ -1,26 +1,111 @@
-= Register Extensions
+= Register Processors
```
There are 9 choices between the processor options and the group options. You are not saying which  you refer to, but I'd guess you are suggesting 2 of the 3 choices for processor and a block for the group.
The block-example.adoc page has 3 of the 9 possibilities, class/class, block/block, and object/object.  I think it points out the differences to some extent, and in particular shows uses for the Object choices.
At the moment, I don't understand the motivation behind your rule of thumb, so I can't explain it.

djencks commented on docs/modules/extensions/pages/processor.adoc:
```diff
@@ -0,0 +1,71 @@
+Applies a complicated algorithm to decode a single array or hash into positional attributes and default attributes.
+Overridden in macro processors.
+
+== What can the `process` method do?
```
I disagree.  The return type of the process method is not documented anywhere else and if you are researching how to write something having the information in one place as well as with each processor description is really valuable.  Having this section here would have saved me hours and hours of work writing extensions.

djencks commented on docs/modules/extensions/pages/packaging.adoc:
```diff
@@ -0,0 +1,66 @@
+For instance:
+
+....
+lib/
```
I think I understand what you mean... please check.

djencks commented:
I've responded to most or possibly all of your comments in one way or another, in this branch and ...
* New branch issue-3884-extension-documentation-one-commit with all the doc changes squashed to one commit
* New branch issue-3884-extension-documentation-as-new where I deleted the old docs and added the new ones as one commit

(these might be a lot easier to review than the issue-3884-extension-documentation branch. All 3 branches should have the same content.  I can open PRs for these if desirable, although it seemed like clutter.)

* After some thought, I suggest we keep the nestable-block extension as the triple example in block-example.adoc.  The point of this page is to demonstrate 3 of the 9 different ways of getting the code that does something to be a working extension.  Having the simplest possible extension (this one has a lot fewer LOC than shout block) is ideal for this, I think.  It also gives an opportunity to show a use for objects for groups and processors.

* For the examples, the "out" files  are generated by a script.  If you change something and rerun the script, you can see if something broke with git status.  I think this is better than generating the "out" files during the build.  After symlinks are in Antora we could move all this to the test suite, but I think that's silly.

* I don't understand the desired relationship between the directory structure of a gem and the class structure.  Following your comment I changed the suggested gem directory structure for the nestable example, but I don't understand if it's appropriate to also nest the class as Asciidoctor::Nestable or Asciidoctor::MyExample::Nestable or ???.  Should this affect the examples as well, in which all processors are classes?
